lib/auth/accountrecovery.go

```diff
@@ -33,13 +37,118 @@ import (
 // This is a fake hash used to mitigate timing attacks against invalid usernames or if user does
 // exist but does not have recovery codes.
 var fakeRecoveryCodeHash = []byte(`$2a$10$c2.h4pF9AA25lbrWo6U0D.ZmnYpFDaNzN3weNNYNC3jAkYEX9kpzu`)
 
+// ErrMaxFailedAttemptsFromStartRecovery is a user friendly error message to try again later.
+// This error is defined in a variable so that the root caller can determine if an email needs to be sent.
+var ErrMaxFailedAttemptsFromStartRecovery = trace.AccessDenied(startRecoveryMaxFailedAttemptsErrMsg)
+
+// StartAccountRecovery implements AuthService.StartAccountRecovery.
+func (s *Server) StartAccountRecovery(ctx context.Context, req *proto.StartAccountRecoveryRequest) (types.UserToken, error) {
+	if err := s.isAccountRecoveryAllowed(ctx); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Only user's with email as their username can start recovery.
+	if _, err := mail.ParseAddress(req.GetUsername()); err != nil {
+		log.Debugf("Failed to start account recovery, user %s is not in valid email format", req.GetUsername())
+		return nil, trace.AccessDenied(startRecoveryGenericErrMsg)
+	}
+
+	if err := s.verifyCodeWithRecoveryLock(ctx, req.GetUsername(), req.GetRecoveryCode()); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Remove any other existing tokens for this user before creating a token.
+	if err := s.deleteUserTokens(ctx, req.Username); err != nil {
+		log.Error(trace.DebugReport(err))
+		return nil, trace.AccessDenied(startRecoveryGenericErrMsg)
+	}
+
+	token, err := s.createRecoveryToken(ctx, req.GetUsername(), UserTokenTypeRecoveryStart, req.GetRecoverType())
+	if err != nil {
+		log.Error(trace.DebugReport(err))
+		return nil, trace.AccessDenied(startRecoveryGenericErrMsg)
+	}
+
+	return token, nil
+}
+
+// verifyCodeWithRecoveryLock counts number of failed attempts at providing a valid recovery code.
+// After MaxAccountRecoveryAttempts, user is temporarily locked from further attempts at recovering and also
+// locked from logging in. Modeled after existing function WithUserLock.
+func (s *Server) verifyCodeWithRecoveryLock(ctx context.Context, username string, recoveryCode []byte) error {
+	user, err := s.Identity.GetUser(username, false)
+	switch {
+	case trace.IsNotFound(err):
+		// If user is not found, still authenticate. It should always return an error.
+		// This prevents username oracles and timing attacks.
+		return s.verifyRecoveryCode(ctx, username, recoveryCode)
+	case err != nil:
+		log.Error(trace.DebugReport(err))
+		return trace.AccessDenied(startRecoveryGenericErrMsg)
+	}
+
+	status := user.GetStatus()
+	if status.IsLocked && status.RecoveryAttemptLockExpires.After(s.clock.Now().UTC()) {
+		log.Debugf("%v exceeds %v failed account recovery attempts, locked until %v",
+			user.GetName(), defaults.MaxAccountRecoveryAttempts, apiutils.HumanTimeFormat(status.RecoveryAttemptLockExpires))
+		return trace.AccessDenied(startRecoveryMaxFailedAttemptsErrMsg)
```

Comment:
Should this be:

```suggestion
		return trace.AccessDenied(ErrMaxFailedAttemptsFromStartRecovery)
```

Reply:
no, the `ErrMax...` is an error that will be checked by root caller to email once